docs/notebooks/create_synthetic_data_from_dgan_api.ipynb

```diff
@@ -0,0 +1,535 @@
+        "print(\"Difference in real correlations and synethic data correlations:\")\n",
+        "correlation_heatmap(df[COLUMNS].corr() - synthetic_df[COLUMNS].corr())"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "0FmWGdwkgmlD",
+      "metadata": {
+        "id": "0FmWGdwkgmlD"
+      },
+      "outputs": [],
+      "source": [
+        "# Plot histograms of the distribution of values within each column\n",
+        "h1 = histogram(df['WTI Price'], synthetic_df['WTI Price'])\n",
+        "h2 = histogram(df['Brent Price'], synthetic_df['Brent Price'])\n",
+        "combine_subplots(\n",
+        "    figures=[h1, h2],\n",
+        "    titles=['WTI Price', 'Brent Price'],\n",
+        "    subplot_type = \"xy\",\n",
+        "    shared_xaxes=True,\n",
+        "    shared_yaxes=True,\n",
+        ")\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "095ed91d",
+      "metadata": {
+        "id": "095ed91d"
+      },
+      "outputs": [],
+      "source": [
+        "# Functions to calculate autocorrelation\n",
+        "def autocorr(X, Y):\n",
+        "    EPS = 1e-8\n",
+        "    Xm = torch.mean(X, 1).unsqueeze(1)\n",
+        "    Ym = torch.mean(Y, 1).unsqueeze(1)\n",
+        "    r_num = torch.sum((X - Xm) * (Y - Ym), 1)\n",
+        "    r_den = torch.sqrt(torch.sum((X - Xm)**2, 1) * torch.sum((Y - Ym)**2, 1))\n",
+        "\n",
+        "    r_num[r_num == 0] = EPS\n",
+        "    r_den[r_den == 0] = EPS\n",
+        "\n",
+        "    r = r_num / r_den\n",
+        "    r[r > 1] = 0\n",
+        "    r[r < -1] = 0\n",
+        "\n",
+        "    return r\n",
+        "    \n",
+        "def get_autocorr(feature):\n",
+        "    feature = torch.from_numpy(feature)\n",
+        "    feature_length = feature.shape[1]\n",
+        "    autocorr_vec = torch.Tensor(feature_length-2)\n",
+        "\n",
+        "    for j in range(1, feature_length - 1):\n",
+        "      autocorr_vec[j - 1] = torch.mean(autocorr(feature[:, :-j], feature[:, j:]))\n",
+        "\n",
+        "    return autocorr_vec.cpu().detach().numpy()\n",
+        "\n",
+        "def generate_numpy_for_autocorr(df, batch_size):\n",
+        "    features = df[COLUMNS].to_numpy()\n",
+        "    n = features.shape[0] // batch_size\n",
+        "\n",
+        "    # Shape is now (# examples, # time points, # features)\n",
+        "    features = features[:(n*batch_size),:].reshape(-1, batch_size, features.shape[1])\n",
+        "    return features"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "9bb282d0",
+      "metadata": {
+        "id": "9bb282d0"
+      },
+      "outputs": [],
+      "source": [
+        "# Generate autocorrelations from synthetic and real data and plot\n",
+        "\n",
+        "synthetic_acf = get_autocorr(generate_numpy_for_autocorr(df, MAX_SEQUENCE_LEN))\n",
```

Comment:
These names are flipped, so the legend in the plot is reversed I think. Unless the df names are also the reverse of what's actually in them.

`synthetic_acf` comes from `df` and `acf` comes from `synthetic_df`.